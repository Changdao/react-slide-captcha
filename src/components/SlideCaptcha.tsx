/// <reference path='../../types/global.d.ts'/>

import * as React from 'react';
import './styles/index.less';

enum validateStatus{
  init = 0,
  success = 1,
  error = -1,
}

interface IProps {
  readonly puzzleUrl: string;
  readonly bgUrl: string;
  readonly onRequest: (validateValue: number, validatedSuccess: any, validatedFail?: any) => void;
  readonly slidedImage?: any
  readonly slidedImageSuccess?: any
  readonly slidedImageError?: any
  readonly containerClassName?: string;
}

interface IState {
  originX: number;
  offsetX: number;
  puzzleUrl: string;
  bgUrl: string;
  validated: validateStatus;
  isMoving: boolean;
  isTouchEndSpan: boolean;
}

class SlideCaptcha extends React.Component<IProps, IState>{
  state: IState = {
    originX: 0,
    offsetX: 0,
    puzzleUrl: '',
    bgUrl: '',
    validated: validateStatus.init,
    isMoving: false,
    isTouchEndSpan: false,
  };
  constructor(props: IProps) {
    super(props);
  }

  componentDidMount() {
    setTimeout(() => {
      this.maxSlidedWidth = this.ctrlWidth.clientWidth - this.sliderWidth.clientWidth;
    }, 0);
  }

  private maxSlidedWidth: number = 0;
  private ctrlWidth: any = null;
  private sliderWidth: any = null;

  private getClientX = (e): number => {
    if (e.type.indexOf('mouse') > -1) {
      return e.clientX;
    }
    if (e.type.indexOf('touch') > -1) {
      return e.touches[0].clientX;
    }
  };

  private move = (e): void => {
    const clientX = this.getClientX(e);
    let offsetX = clientX - this.state.originX;
    if (offsetX > 0) {
      if (offsetX > this.maxSlidedWidth) {
        // 超过最大移动范围，按极限值算
        offsetX = this.maxSlidedWidth;
      }
      this.setState({
        offsetX,
        isMoving: true,
      })
    }
  };

  public validatedSuccess = (callback: () => any):void => {
    this.setState({
      validated: validateStatus.success
    },callback());

  };

  public validatedFail = (callback: () => any): any => {
    this.setState({
      validated: validateStatus.error
    },callback());
  };

  private handleTouchStart = (e): void => {
    e.preventDefault();
    this.setState({
      originX: this.getClientX(e),
    })
  };

  private handleTouchMove = (e): void => {
    e.preventDefault();
    this.move(e);
  };

  private handleTouchEnd = (e): void => {
    e.preventDefault();
    if (this.state.offsetX > 0) {
      const validateValue = this.state.offsetX / this.maxSlidedWidth;
      this.setState({
        isTouchEndSpan: true,
        isMoving: false
      });
      if(this.props.onRequest) {
        this.props.onRequest(validateValue, this.validatedSuccess, this.validatedFail);
      }
      setTimeout(() => {
        this.setState({
          offsetX: 0,
          originX: 0,
          isTouchEndSpan: false,
          isMoving: false,
          validated: validateStatus.init,
        })
      }, 500);
    } else {
      this.setState({
        isTouchEndSpan: false,
        isMoving: false,
        offsetX: 0,
        originX: 0,
        validated: validateStatus.init,
      });
    }
  };

  renderImage = ():any =>{
    let slidedImageValue = this.props.slidedImage || '>';
    let slidedImageSuccessValue = this.props.slidedImageSuccess || '>';
    let slidedImageErrorValue = this.props.slidedImageError || 'x';
    return { slidedImageValue, slidedImageSuccessValue, slidedImageErrorValue }
  };

  renderCtrlClassName = (slidedImage, slidedImageSuccess, slidedImageError) => {
    let ctrlClassName;
    let slidedImageValue = slidedImage;
    if(this.state.isMoving) {
      ctrlClassName = 'slider-moving';
    } else {
      if(this.state.isTouchEndSpan) {
        if(this.state.validated === validateStatus.success) {
          ctrlClassName = 'slider-end slider-success';
          slidedImageValue = slidedImageSuccess;
        } else if(this.state.validated === validateStatus.error) {
          ctrlClassName = 'slider-end slider-error';
          slidedImageValue = slidedImageError;
        } else {
          ctrlClassName = 'slider-moving';
        }
      } else {
        ctrlClassName = '';
      }
    }
    return { ctrlClassName, slidedImage: slidedImageValue };
  };

  render() {
    const {
      slidedImageValue, slidedImageSuccessValue, slidedImageErrorValue
    } = this.renderImage();

    let { ctrlClassName, slidedImage } = this.renderCtrlClassName(slidedImageValue, slidedImageSuccessValue, slidedImageErrorValue);
    return(
      <div>
        <div className={`slideCaptchaContainer ${this.props.containerClassName ? this.props.containerClassName: ''}`}>
          <div className="panel">
            <img src={this.props.bgUrl} className="bgImg" />
            <img src={this.props.puzzleUrl} className="puzzleImg"  style={{left: `${this.state.offsetX}px`}} />
          </div>
          <div className={`control ${ctrlClassName ? ctrlClassName: ''}`} ref={(el)=>  { this.ctrlWidth = el; } }>
            <div className="slided" style={{width: `${this.state.offsetX}px`}} />
            <div className="slider"
                 ref={(el) => { this.sliderWidth = el;}}
                 style={{left: `${this.state.offsetX}px`}}
                 onTouchStart={this.handleTouchStart}
                 onTouchMove={this.handleTouchMove}
                 onTouchEnd={this.handleTouchEnd}
            >
              {slidedImage}
            </div>
            <div className="tips">
              <span>
                向右滑动滑块填充拼图
              </span>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export default SlideCaptcha;
