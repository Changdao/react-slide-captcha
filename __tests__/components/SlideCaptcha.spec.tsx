import * as React from 'react';
import { shallow } from 'enzyme';
import SlideCaptcha from '../../src/index';

test('SlideCaptcha', () => {
  const component = shallow(<SlideCaptcha
    puzzleUrl="www.baidu.com"
    bgUrl="www.baidu.com"
    onRequest={() => console.log(123)}
    containerClassName="test"
    tipsText="请向右滑动滑块填充拼图"
    style={{width: '500px'}}
  />);

  expect(component).toMatchSnapshot();

});
