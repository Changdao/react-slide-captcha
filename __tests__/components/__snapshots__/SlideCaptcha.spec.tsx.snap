// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlideCaptcha 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SlideCaptcha
    bgUrl="www.baidu.com"
    containerClassName="test"
    onRequest={[Function]}
    puzzleUrl="www.baidu.com"
    style={
      Object {
        "width": "500px",
      }
    }
    tipsText="请向右滑动滑块填充拼图"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="slideCaptchaContainer test"
        onMouseMove={[Function]}
        onMouseUp={[Function]}
        style={
          Object {
            "width": "500px",
          }
        }
      >
        <div
          className="panel"
        >
          <img
            className="bgImg"
            src="www.baidu.com"
          />
          <img
            className="puzzleImg"
            src="www.baidu.com"
            style={
              Object {
                "left": "0px",
              }
            }
          />
        </div>
        <div
          className="control "
        >
          <div
            className="slided"
            style={
              Object {
                "width": "0px",
              }
            }
          />
          <div
            className="slider"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "left": "0px",
              }
            }
          >
            &gt;
          </div>
          <div
            className="tips"
          >
            <span>
              请向右滑动滑块填充拼图
            </span>
          </div>
        </div>
      </div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="panel"
          >
            <img
              className="bgImg"
              src="www.baidu.com"
            />
            <img
              className="puzzleImg"
              src="www.baidu.com"
              style={
                Object {
                  "left": "0px",
                }
              }
            />
          </div>,
          <div
            className="control "
          >
            <div
              className="slided"
              style={
                Object {
                  "width": "0px",
                }
              }
            />
            <div
              className="slider"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              style={
                Object {
                  "left": "0px",
                }
              }
            >
              &gt;
            </div>
            <div
              className="tips"
            >
              <span>
                请向右滑动滑块填充拼图
              </span>
            </div>
          </div>,
        ],
        "className": "slideCaptchaContainer test",
        "onMouseMove": [Function],
        "onMouseUp": [Function],
        "style": Object {
          "width": "500px",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <img
                className="bgImg"
                src="www.baidu.com"
              />,
              <img
                className="puzzleImg"
                src="www.baidu.com"
                style={
                  Object {
                    "left": "0px",
                  }
                }
              />,
            ],
            "className": "panel",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "bgImg",
                "src": "www.baidu.com",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "puzzleImg",
                "src": "www.baidu.com",
                "style": Object {
                  "left": "0px",
                },
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="slided"
                style={
                  Object {
                    "width": "0px",
                  }
                }
              />,
              <div
                className="slider"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                style={
                  Object {
                    "left": "0px",
                  }
                }
              >
                &gt;
              </div>,
              <div
                className="tips"
              >
                <span>
                  请向右滑动滑块填充拼图
                </span>
              </div>,
            ],
            "className": "control ",
          },
          "ref": [Function],
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "slided",
                "style": Object {
                  "width": "0px",
                },
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": ">",
                "className": "slider",
                "onMouseDown": [Function],
                "onTouchEnd": [Function],
                "onTouchMove": [Function],
                "onTouchStart": [Function],
                "style": Object {
                  "left": "0px",
                },
              },
              "ref": [Function],
              "rendered": ">",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <span>
                  请向右滑动滑块填充拼图
                </span>,
                "className": "tips",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "请向右滑动滑块填充拼图",
                },
                "ref": null,
                "rendered": "请向右滑动滑块填充拼图",
                "type": "span",
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="slideCaptchaContainer test"
          onMouseMove={[Function]}
          onMouseUp={[Function]}
          style={
            Object {
              "width": "500px",
            }
          }
        >
          <div
            className="panel"
          >
            <img
              className="bgImg"
              src="www.baidu.com"
            />
            <img
              className="puzzleImg"
              src="www.baidu.com"
              style={
                Object {
                  "left": "0px",
                }
              }
            />
          </div>
          <div
            className="control "
          >
            <div
              className="slided"
              style={
                Object {
                  "width": "0px",
                }
              }
            />
            <div
              className="slider"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              style={
                Object {
                  "left": "0px",
                }
              }
            >
              &gt;
            </div>
            <div
              className="tips"
            >
              <span>
                请向右滑动滑块填充拼图
              </span>
            </div>
          </div>
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="panel"
            >
              <img
                className="bgImg"
                src="www.baidu.com"
              />
              <img
                className="puzzleImg"
                src="www.baidu.com"
                style={
                  Object {
                    "left": "0px",
                  }
                }
              />
            </div>,
            <div
              className="control "
            >
              <div
                className="slided"
                style={
                  Object {
                    "width": "0px",
                  }
                }
              />
              <div
                className="slider"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                style={
                  Object {
                    "left": "0px",
                  }
                }
              >
                &gt;
              </div>
              <div
                className="tips"
              >
                <span>
                  请向右滑动滑块填充拼图
                </span>
              </div>
            </div>,
          ],
          "className": "slideCaptchaContainer test",
          "onMouseMove": [Function],
          "onMouseUp": [Function],
          "style": Object {
            "width": "500px",
          },
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <img
                  className="bgImg"
                  src="www.baidu.com"
                />,
                <img
                  className="puzzleImg"
                  src="www.baidu.com"
                  style={
                    Object {
                      "left": "0px",
                    }
                  }
                />,
              ],
              "className": "panel",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "bgImg",
                  "src": "www.baidu.com",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "puzzleImg",
                  "src": "www.baidu.com",
                  "style": Object {
                    "left": "0px",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="slided"
                  style={
                    Object {
                      "width": "0px",
                    }
                  }
                />,
                <div
                  className="slider"
                  onMouseDown={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  style={
                    Object {
                      "left": "0px",
                    }
                  }
                >
                  &gt;
                </div>,
                <div
                  className="tips"
                >
                  <span>
                    请向右滑动滑块填充拼图
                  </span>
                </div>,
              ],
              "className": "control ",
            },
            "ref": [Function],
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "slided",
                  "style": Object {
                    "width": "0px",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": ">",
                  "className": "slider",
                  "onMouseDown": [Function],
                  "onTouchEnd": [Function],
                  "onTouchMove": [Function],
                  "onTouchStart": [Function],
                  "style": Object {
                    "left": "0px",
                  },
                },
                "ref": [Function],
                "rendered": ">",
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <span>
                    请向右滑动滑块填充拼图
                  </span>,
                  "className": "tips",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "请向右滑动滑块填充拼图",
                  },
                  "ref": null,
                  "rendered": "请向右滑动滑块填充拼图",
                  "type": "span",
                },
                "type": "div",
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
